core/src/graph/key_manager.rs

```diff
@@ -0,0 +1,281 @@
+use crate::{
+	dsnp::{
+		api_types::{DsnpKeys, KeyData, PageHash},
+		dsnp_types::{DsnpPublicKey, DsnpUserId},
+		reader_writer::{DsnpReader, DsnpWriter},
+	},
+	frequency::Frequency,
+};
+use anyhow::{Error, Result};
+use dryoc::keypair::StackKeyPair;
+use std::{cell::RefCell, collections::HashMap, ops::Deref, rc::Rc, sync::RwLock};
+
+trait PublicKeyProvider {
+	fn import_dsnp_keys(&mut self, keys: DsnpKeys) -> Result<()>;
+	fn export_new_keys(&self) -> Result<Vec<DsnpKeys>>;
+	fn get_all_keys(&self, dsnp_user_id: DsnpUserId) -> Vec<&DsnpPublicKey>;
+	fn get_key_by_id(&self, dsnp_user_id: DsnpUserId, key_id: u64) -> Option<&DsnpPublicKey>;
+	fn get_key_by_public_key(
+		&self,
+		dsnp_user_id: DsnpUserId,
+		public_key: Vec<u8>,
+	) -> Option<&DsnpPublicKey>;
+	fn get_active_key(&self, dsnp_user_id: DsnpUserId) -> Option<&DsnpPublicKey>;
+	fn add_new_key(&mut self, dsnp_user_id: DsnpUserId, public_key: Vec<u8>) -> Result<u64>;
+}
+
+trait UserKeyProvider {
+	fn import_key_pairs(&mut self, pairs: Vec<StackKeyPair>);
+	fn get_key(&self, key_id: u64) -> Option<(DsnpPublicKey, StackKeyPair)>;
+}
+
+pub struct PublicKeyManager {
+	/// keys are stored sorted by index
+	dsnp_user_to_keys: HashMap<DsnpUserId, (Vec<DsnpPublicKey>, PageHash)>,
+	new_keys: HashMap<DsnpUserId, DsnpPublicKey>,
+}
+
+pub struct UserKeyManager {
+	public_key_manager: RwLock<PublicKeyManager>,
```

Comment:
This is giving me a hard time. I tried a Rc and RefCell and .. but no success. I guess I have to reread all those.

The idea is to have one instance of PublicKeyManager and share that reference with UserKeyManagers. UserKeyManager should be only reading from it.